graphbolt/src/cuda/extension/unique_and_compact_map.cu

```diff
@@ -0,0 +1,256 @@
+      const int64_t tensor_index = indexes[i];
+      const auto tensor_offset = i - offsets[tensor_index];
+      const int64_t node_id = pointers[tensor_index][tensor_offset];
+      const auto batch_index = tensor_index / 2;
+      const int64_t key = node_id | (batch_index << node_id_bits);
+
+      auto slot = map.find(key);
+      const auto batch_offset = offsets[batch_index * 2];
+      const auto new_id = valid_i - valid[batch_offset];
+      unique_ids[valid_i] = node_id;
+      slot->second = new_id;
+    }
+
+    i += stride;
+  }
+}
+
+template <typename index_t, typename map_t>
+__global__ void _MapIdsBatched(
+    const int num_batches, const int64_t num_edges,
+    const int32_t* const indexes, index_t** pointers,
+    const int64_t* const offsets, map_t map, index_t* mapped_ids) {
+  int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  const int stride = gridDim.x * blockDim.x;
+
+  while (i < num_edges) {
+    const int64_t tensor_index = indexes[i];
+    int64_t batch_index;
+
+    if (tensor_index >= 2 * num_batches) {
+      batch_index = tensor_index - 2 * num_batches;
+    } else if (tensor_index & 1) {
+      batch_index = tensor_index / 2;
+    } else {
+      batch_index = -1;
+    }
+
+    // Only map src or dst ids.
+    if (batch_index >= 0) {
+      const auto tensor_offset = i - offsets[tensor_index];
+      const int64_t node_id = pointers[tensor_index][tensor_offset];
+      const int64_t key = node_id | (batch_index << node_id_bits);
+
+      auto slot = map.find(key);
+      mapped_ids[i] = slot->second;
+    }
+
+    i += stride;
+  }
+}
+
+std::vector<std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> >
+UniqueAndCompactBatchedMap(
+    const std::vector<torch::Tensor>& src_ids,
+    const std::vector<torch::Tensor>& dst_ids,
+    const std::vector<torch::Tensor>& unique_dst_ids) {
+  auto allocator = cuda::GetAllocator();
+  auto stream = cuda::GetCurrentStream();
+  auto scalar_type = src_ids.at(0).scalar_type();
+  constexpr int BLOCK_SIZE = 512;
+  const auto num_batches = src_ids.size();
+  static_assert(
+      sizeof(std::ptrdiff_t) == sizeof(int64_t),
+      "Need to be compiled on a 64-bit system.");
+  constexpr int batch_id_bits = sizeof(int64_t) * 8 - 1 - node_id_bits;
+  TORCH_CHECK(
+      num_batches <= (1 << batch_id_bits),
+      "UniqueAndCompactBatched supports a batch size of up to ",
+      1 << batch_id_bits);
+  return AT_DISPATCH_INDEX_TYPES(
+      scalar_type, "unique_and_compact", ([&] {
+        auto pointers_and_offsets = torch::empty(
+            6 * num_batches + 1,
+            c10::TensorOptions().dtype(torch::kInt64).pinned_memory(true));
+        auto pointers_ptr =
+            reinterpret_cast<index_t**>(pointers_and_offsets.data_ptr());
+        auto offsets_ptr =
+            pointers_and_offsets.data_ptr<int64_t>() + 3 * num_batches;
+        for (std::size_t i = 0; i < num_batches; i++) {
+          pointers_ptr[2 * i] = unique_dst_ids.at(i).data_ptr<index_t>();
```

Comment:
It looks it request a chunk of memory and split it logically to different memory area, but problem is the naming is really confusing, can you replace it with a more meaningful name?
Likely, what it points for `pointers_ptr`, what offset is for `offset_ptr`.

Reply:
I added comments to explain how the allocated storage is used.